### ISSUE TITLE
Add support for tail-based sampling

### ISSUE THREAD

**Issue by bcharna** [getsentry/sentry-ruby#1528]
**Describe the idea**
Add support for tail-based sampling.

**Why do you think it's beneficial to most of the users**
When using tail-based sampling, you are able to ensure that all transactions/traces with errors are sent to Sentry. This is useful, especially when your service's traffic volume is on the low end but you want _all_ traces with errors reported. Note that this could get a bit expensive to do in-line, but shouldn't be an issue for services with low traffic. A further enhancement would be to have a separate collector service to do the processing outside of the service being instrumented (an example of this is Honeycomb's [Refinery](https://github.com/honeycombio/refinery) or using the [gateway pattern](https://opentelemetry.io/docs/collector/getting-started/#gateway) with the OpenTelemetry Collector).

**Comment by st0012** [comment]
I've proposed something similar: https://github.com/getsentry/sentry-ruby/issues/1470 🙂 
TL;DR: Sentry is trying to simplify the trace sampling mechanism. But because it's still a WIP, the SDK side won't make changes until the plan is clear.

**Comment by st0012** [comment]
@bcharna after version `4.6.5`, you should be able to achieve tail-based filtering with event processors. for example, this snippet drops transactions with `not_found` status.

```ruby
Sentry.configure_scope do |scope|
  scope.add_event_processor do |event, hint|
    if event.type == "transaction" && event.contexts.dig(:trace, :status) == "not_found"
      nil
    else
      event
    end
  end
end
```

so I'm closing this now.

**Comment by bcharna** [comment]
@st0012 noted regarding Sentry currently working on the trace sampling mechanism. Your workaround is interesting and while you didn't imply that it's a perfect replacement for tail-based sampling, I just want to point out one core issue with this approach: In an ideal tail-based sampling setup, either all spans across the services should be collected, or none should be. Since event processors apply to one transaction in isolation , this all or nothing guarantee can't be achieved.

Sort of an aside – it would be great if Sentry could display analytics/charts more correctly based on sampled data. For example, if a particular transaction is sampled at 50%, when looking at a chart using a `count` query in Discover, Sentry would double the amount of traces it receives in charts. Honeycomb does this by allowing you to send sample rate for each trace, see [here](https://docs.honeycomb.io/manage-data-volume/sampling/#calculating-with-sampled-data).
